src/AspNetCoreModuleV2/CommonLib/FileOutputManager.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+#include "stdafx.h"
+
+FileOutputManager::~FileOutputManager()
+{
+    HANDLE   handle = NULL;
+    WIN32_FIND_DATA fileData;
+
+    if (m_hLogFileHandle != INVALID_HANDLE_VALUE)
+    {
+        m_Timer.CancelTimer();
+        CloseHandle(m_hLogFileHandle);
+        m_hLogFileHandle = INVALID_HANDLE_VALUE;
+    }
+
+    // delete empty log file
+    handle = FindFirstFile(m_struLogFilePath.QueryStr(), &fileData);
+    if (handle != INVALID_HANDLE_VALUE &&
+        fileData.nFileSizeHigh == 0 &&
+        fileData.nFileSizeLow == 0) // skip check of nFileSizeHigh
+    {
+        FindClose(handle);
+        // no need to check whether the deletion succeeds
+        // as nothing can be done
+        DeleteFile(m_struLogFilePath.QueryStr());
+    }
+
+    _dup2(m_fdStdOut, _fileno(stdout));
+    _dup2(m_fdStdErr, _fileno(stderr));
+}
+
+void FileOutputManager::NotifyStartupComplete()
+{
+}
+
+bool FileOutputManager::GetStdOutContent(STRU* struStdOutput)
+{
+    //
+    // Ungraceful shutdown, try to log an error message.
+    // This will be a common place for errors as it means the hostfxr_main returned
+    // or there was an exception.
+    //
+    CHAR            pzFileContents[4096] = { 0 };
+    DWORD           dwNumBytesRead;
+    LARGE_INTEGER   li = { 0 };
+    BOOL            fLogged = FALSE;
+    DWORD           dwFilePointer = 0;
+
+    if (m_hLogFileHandle != INVALID_HANDLE_VALUE)
+    {
+        if (GetFileSizeEx(m_hLogFileHandle, &li) && li.LowPart > 0 && li.HighPart == 0)
+        {
+            if (li.LowPart > 4096)
+            {
+                dwFilePointer = SetFilePointer(m_hLogFileHandle, -4096, NULL, FILE_END);
+            }
+            else
+            {
+                dwFilePointer = SetFilePointer(m_hLogFileHandle, 0, NULL, FILE_BEGIN);
+            }
+
+            if (dwFilePointer != INVALID_SET_FILE_POINTER)
+            {
+                // Read file fails.
+                if (ReadFile(m_hLogFileHandle, pzFileContents, 4096, &dwNumBytesRead, NULL))
+                {
+                    if (SUCCEEDED(struStdOutput->CopyA(pzFileContents, dwNumBytesRead)))
+                    {
+                        fLogged = TRUE;
+                    }
+                }
+            }
+        }
+    }
+
+    return fLogged;
+}
+
+FileOutputManager::FileOutputManager(PCWSTR pwzStdOutLogFileName, PCWSTR pwzApplicationPath)
+    : m_hLogFileHandle(INVALID_HANDLE_VALUE)
+{
+    m_struStdOutLogFileName.Copy(pwzStdOutLogFileName);
```

Comment:
Copy may fail, please check the returned hr and you don't want to fail the constructor. Please add an Initialize function.

Reply:
I'll just use wstrings for these instead. I'd rather not add an Initalize function just for that HRESULTs.